airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.kt

```diff
@@ -460,6 +478,99 @@ open class S3StorageOperations(
         }
     }
 
+    override fun listExistingObjects(
+        namespace: String?,
+        streamName: String,
+        objectPath: String,
+        pathFormat: String,
+        currentGenerationId: Long
+    ): List<String> {
+        val regexFormat: Pattern =
+            Pattern.compile(getRegexFormat(namespace, streamName, pathFormat))
+        var keys = listOf<S3ObjectSummary>()
+        var objects: ObjectListing = listObjects(objectPath)
+        while (objects.objectSummaries.size > 0) {
+            keys =
+                keys +
+                    objects.objectSummaries.filter { obj: S3ObjectSummary ->
+                        regexFormat
+                            .matcher(
+                                obj.key,
+                            )
+                            .matches()
+                    }
+            if (objects.isTruncated) {
+                objects = s3Client.listNextBatchOfObjects(objects)
+            } else {
+                break
+            }
+        }
+        if (currentGenerationId != -1L) {
+            keys = sliceObjectsByMetadata(keys, currentGenerationId)
+        }
+
+        return keys.map { obj: S3ObjectSummary -> obj.key }
+    }
+
+    private fun sliceObjectsByMetadata(
```

Comment:
nit: naming is not super clear, maybe `findOldGenerationObjects` + rename generationId to `minGeneration`?

Reply:
renamed to 
```
private fun filterObjectsByGenerationLessThan(
        keyList: List<S3ObjectSummary>,
        maxGenerationId: Long
    ):
```